arangod/Agency/Job.cpp

```diff
@@ -464,13 +466,30 @@ std::string Job::findNonblockedCommonHealthyInSyncFollower(  // Which is in "GOO
       continue;
     }  // if
 
-    for (const auto& server :
-         VPackArrayIterator(snap.hasAsArray(currentShardPath).first)) {
-      auto id = server.copyString();
+    bool isArray = false;
+    VPackSlice serverList;
+    // If we do have failover candidates, we should use them
+    std::tie(serverList, isArray) = snap.hasAsArray(currentFailoverCandidatesPath);
+    if (!isArray) {
+      // We have old DBServers that do not report failover candidates,
+      // Need to rely on current
+      std::tie(serverList, isArray) = snap.hasAsArray(currentShardPath);
+      TRI_ASSERT(isArray);
+      if (!isArray) {
+        THROW_ARANGO_EXCEPTION_MESSAGE(
```

Comment:
I have no clue if it is a good idea to throw an exception here. Nobody knows where it is catched. Could be, that the current supervision run terminated and thus the supervision will never make progress.

Reply:
Same for me. But I'd prefer that we make sure exceptions are handled sanely and leave this here if possible.

Reply:
I read the code and traced it down Is I have been on the same page as you.
Before:
The serverList (as notArray) was handed into a VPackArrayIterator, this throws an error (Not an array).
I now throw this error earlier.

The error is essentially only catched on the most outer side with a catch all. (3 or 4 levels up)
As I did not intend to refactor such large parts of the Agency code i just keep it this way, it is not worse than before.